File: app/src/main/java/knf/kuma/explorer/FragmentFiles.java
package knf.kuma.explorer;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.ExplorerObject;

public class FragmentFiles extends Fragment {

    public static final String TAG = "Files";
    @BindView(R.id.recycler)
    RecyclerView recyclerView;
    @BindView(R.id.error)
    View error;
    @BindView(R.id.progress)
    ProgressBar progressBar;
    @BindView(R.id.state)
    TextView state;
    private SelectedListener listener;
    private ExplorerFilesAdapter adapter;
    private boolean isFist = true;

    private int count = 0;

    public FragmentFiles() {
    }

    public static FragmentFiles get(SelectedListener listener) {
        FragmentFiles fragmentFiles = new FragmentFiles();
        fragmentFiles.setListener(listener);
        return fragmentFiles;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        CacheDB.INSTANCE.explorerDAO().getAll().observe(this, explorerObjects -> {
            adapter.update(explorerObjects);
            if (explorerObjects.size() != 0) {
                progressBar.setVisibility(View.GONE);
                state.setVisibility(View.GONE);
                if (isFist) {
                    isFist = false;
                    recyclerView.scheduleLayoutAnimation();
                }
            }
        });
        ExplorerCreator.getStateListener().observe(this, s -> {
            state.setText(s);
            state.setVisibility(s == null ? View.GONE : View.VISIBLE);
        });
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayout(), container, false);
        ButterKnife.bind(this, view);
        adapter = new ExplorerFilesAdapter(this, listener);
        recyclerView.setAdapter(adapter);
        return view;
    }

    @LayoutRes
    private int getLayout() {
        if (PreferenceManager.getDefaultSharedPreferences(getContext()).getString("lay_type", "0").equals("0")) {
            return R.layout.recycler_explorer;
        } else {
            return R.layout.recycler_explorer_grid;
        }
    }

    public void onEmpty() {
        new Handler(Looper.getMainLooper()).post(() -> {
            progressBar.setVisibility(View.GONE);
            error.setVisibility(View.VISIBLE);
            state.setVisibility(View.GONE);
        });
    }

    public void setListener(SelectedListener listener) {
        this.listener = listener;
        if (adapter != null)
            adapter.setListener(listener);
    }

    public interface SelectedListener {
        void onSelected(ExplorerObject object);
    }
}

File: app/src/main/java/knf/kuma/explorer/FragmentFiles.kt
package knf.kuma.explorer

import android.os.Bundle
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import knf.kuma.R
import knf.kuma.commons.doOnUI
import knf.kuma.database.CacheDB
import knf.kuma.pojos.ExplorerObject
import org.jetbrains.anko.find

class FragmentFiles : Fragment() {
    lateinit var recyclerView: RecyclerView
    lateinit var error: View
    lateinit var progressBar: ProgressBar
    lateinit var state: TextView
    private var listener: SelectedListener? = null
    private var adapter: ExplorerFilesAdapter? = null
    private var isFist = true

    private val layout: Int
        @LayoutRes
        get() = if (PreferenceManager.getDefaultSharedPreferences(context).getString("lay_type", "0") == "0") {
            R.layout.recycler_explorer
        } else {
            R.layout.recycler_explorer_grid
        }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        CacheDB.INSTANCE.explorerDAO().all.observe(this, Observer { explorerObjects ->
            adapter?.update(explorerObjects)
            if (explorerObjects.isNotEmpty()) {
                progressBar.visibility = View.GONE
                state.visibility = View.GONE
                if (isFist) {
                    isFist = false
                    recyclerView.scheduleLayoutAnimation()
                }
            }
        })
        ExplorerCreator.stateListener.observe(this, Observer { s ->
            state.text = s
            state.visibility = if (s == null) View.GONE else View.VISIBLE
        })
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(layout, container, false)
        recyclerView = view.find(R.id.recycler)
        error = view.find(R.id.error)
        progressBar = view.find(R.id.progress)
        state = view.find(R.id.state)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = ExplorerFilesAdapter(this, listener)
        recyclerView.adapter = adapter
    }

    fun onEmpty() {
        doOnUI {
            progressBar.visibility = View.GONE
            error.visibility = View.VISIBLE
            state.visibility = View.GONE
        }
    }

    fun setListener(listener: SelectedListener) {
        this.listener = listener
        if (adapter != null)
            adapter!!.setListener(listener)
    }

    interface SelectedListener {
        fun onSelected(explorerObject: ExplorerObject)
    }

    companion object {

        const val TAG = "Files"

        operator fun get(listener: SelectedListener): FragmentFiles {
            val fragmentFiles = FragmentFiles()
            fragmentFiles.setListener(listener)
            return fragmentFiles
        }
    }
}
